File: 01_项目源码与运行工程/SafeLab-Vision-Pro/rknn_runtime/yolov8_postprocess.cpp
#include "yolov8_postprocess.hpp"

#include <algorithm>
#include <cmath>
#include <sstream>
#include <stdexcept>

namespace safelab {

const std::vector<std::string>& labels() {
    static const std::vector<std::string> kLabels = {
        "person", "helmet", "vest", "goggles", "gloves", "fire", "smoke"};
    return kLabels;
}

float dequantize_int8(int value, int zero_point, float scale) {
    return static_cast<float>(value - zero_point) * scale;
}

namespace {

float letterbox_scale(const LetterboxMeta& meta) {
    return std::min(
        static_cast<float>(meta.input_width) / static_cast<float>(meta.source_width),
        static_cast<float>(meta.input_height) / static_cast<float>(meta.source_height));
}

float iou_xyxy(const RawDetection& a, const RawDetection& b) {
    const float ix1 = std::max(a.x1, b.x1);
    const float iy1 = std::max(a.y1, b.y1);
    const float ix2 = std::min(a.x2, b.x2);
    const float iy2 = std::min(a.y2, b.y2);
    const float iw = std::max(ix2 - ix1, 0.0f);
    const float ih = std::max(iy2 - iy1, 0.0f);
    const float intersection = iw * ih;
    const float area_a = std::max(a.x2 - a.x1, 0.0f) * std::max(a.y2 - a.y1, 0.0f);
    const float area_b = std::max(b.x2 - b.x1, 0.0f) * std::max(b.y2 - b.y1, 0.0f);
    const float denom = area_a + area_b - intersection;
    if (denom <= 0.0f) {
        return 0.0f;
    }
    return intersection / denom;
}

bool xywh_to_source_xyxy(
    float cx,
    float cy,
    float width,
    float height,
    const LetterboxMeta& meta,
    RawDetection* detection) {
    const float scale = letterbox_scale(meta);
    const float pad_x = (meta.input_width - meta.source_width * scale) / 2.0f;
    const float pad_y = (meta.input_height - meta.source_height * scale) / 2.0f;

    float x1 = (cx - width / 2.0f - pad_x) / scale;
    float y1 = (cy - height / 2.0f - pad_y) / scale;
    float x2 = (cx + width / 2.0f - pad_x) / scale;
    float y2 = (cy + height / 2.0f - pad_y) / scale;

    x1 = std::max(0.0f, std::min(x1, static_cast<float>(meta.source_width - 1)));
    y1 = std::max(0.0f, std::min(y1, static_cast<float>(meta.source_height - 1)));
    x2 = std::max(0.0f, std::min(x2, static_cast<float>(meta.source_width - 1)));
    y2 = std::max(0.0f, std::min(y2, static_cast<float>(meta.source_height - 1)));
    if (x2 <= x1 || y2 <= y1) {
        return false;
    }

    detection->x1 = x1;
    detection->y1 = y1;
    detection->x2 = x2;
    detection->y2 = y2;
    return true;
}

int clip_int(float value, int min_value, int max_value) {
    return std::max(min_value, std::min(static_cast<int>(std::round(value)), max_value));
}

std::string json_escape(const std::string& value) {
    std::ostringstream escaped;
    for (const char ch : value) {
        if (ch == '"' || ch == '\\') {
            escaped << '\\';
        }
        escaped << ch;
    }
    return escaped.str();
}

}  // namespace

std::vector<RawDetection> decode_yolov8_channel_major(
    const std::vector<float>& output,
    int channels,
    int anchors,
    const LetterboxMeta& meta,
    float confidence_threshold,
    float nms_iou_threshold) {
    if (channels <= 4) {
        throw std::runtime_error("YOLOv8 output must have box channels plus at least one class channel");
    }
    if (static_cast<int>(output.size()) != channels * anchors) {
        throw std::runtime_error("YOLOv8 output size does not match channels * anchors");
    }

    std::vector<RawDetection> candidates;
    const int class_count = channels - 4;
    for (int anchor = 0; anchor < anchors; ++anchor) {
        const float cx = output[0 * anchors + anchor];
        const float cy = output[1 * anchors + anchor];
        const float width = output[2 * anchors + anchor];
        const float height = output[3 * anchors + anchor];

        int best_class = 0;
        float best_score = output[4 * anchors + anchor];
        for (int class_id = 1; class_id < class_count; ++class_id) {
            const float score = output[(4 + class_id) * anchors + anchor];
            if (score > best_score) {
                best_score = score;
                best_class = class_id;
            }
        }
        if (best_score < confidence_threshold) {
            continue;
        }

        RawDetection detection{};
        if (!xywh_to_source_xyxy(cx, cy, width, height, meta, &detection)) {
            continue;
        }
        detection.confidence = best_score;
        detection.class_id = best_class;
        candidates.push_back(detection);
    }
    return nms(candidates, nms_iou_threshold);
}

std::vector<RawDetection> nms(const std::vector<RawDetection>& boxes, float iou_threshold) {
    std::vector<RawDetection> remaining = boxes;
    std::sort(remaining.begin(), remaining.end(), [](const RawDetection& a, const RawDetection& b) {
        return a.confidence > b.confidence;
    });

    std::vector<RawDetection> selected;
    while (!remaining.empty()) {
        const RawDetection current = remaining.front();
        selected.push_back(current);
        std::vector<RawDetection> next;
        for (size_t index = 1; index < remaining.size(); ++index) {
            const RawDetection& candidate = remaining[index];
            if (candidate.class_id != current.class_id || iou_xyxy(candidate, current) <= iou_threshold) {
                next.push_back(candidate);
            }
        }
        remaining.swap(next);
    }
    return selected;
}

Detection to_detection(
    const RawDetection& raw,
    int frame_id,
    const std::string& source_type,
    const std::string& model_name,
    float infer_time_ms,
    int frame_width,
    int frame_height) {
    return to_detection(
        raw,
        frame_id,
        source_type,
        model_name,
        infer_time_ms,
        frame_width,
        frame_height,
        labels());
}

Detection to_detection(
    const RawDetection& raw,
    int frame_id,
    const std::string& source_type,
    const std::string& model_name,
    float infer_time_ms,
    int frame_width,
    int frame_height,
    const std::vector<std::string>& label_names) {
    if (raw.class_id < 0 || raw.class_id >= static_cast<int>(label_names.size())) {
        throw std::runtime_error("class_id outside labels range");
    }
    const int x1 = clip_int(raw.x1, 0, frame_width - 1);
    const int y1 = clip_int(raw.y1, 0, frame_height - 1);
    const int x2 = clip_int(raw.x2, 0, frame_width - 1);
    const int y2 = clip_int(raw.y2, 0, frame_height - 1);
    Detection detection{};
    detection.frame_id = frame_id;
    detection.source_type = source_type;
    detection.class_name = label_names[raw.class_id];
    detection.confidence = raw.confidence;
    detection.bbox = {x1, y1, x2, y2};
    detection.center = {(x1 + x2) / 2, (y1 + y2) / 2};
    detection.area = std::max(x2 - x1, 0) * std::max(y2 - y1, 0);
    detection.model_name = model_name;
    detection.infer_time_ms = infer_time_ms;
    return detection;
}

std::string detection_to_json(const Detection& detection) {
    std::ostringstream json;
    json << "{";
    json << "\"frame_id\":" << detection.frame_id << ",";
    json << "\"source_type\":\"" << json_escape(detection.source_type) << "\",";
    json << "\"class_name\":\"" << json_escape(detection.class_name) << "\",";
    json << "\"confidence\":" << detection.confidence << ",";
    json << "\"bbox\":[" << detection.bbox[0] << "," << detection.bbox[1] << ","
         << detection.bbox[2] << "," << detection.bbox[3] << "],";
    json << "\"center\":[" << detection.center[0] << "," << detection.center[1] << "],";
    json << "\"area\":" << detection.area << ",";
    json << "\"model_name\":\"" << json_escape(detection.model_name) << "\",";
    json << "\"infer_time_ms\":" << detection.infer_time_ms;
    json << "}";
    return json.str();
}

}  // namespace safelab

File: 01_项目源码与运行工程/SafeLab-Vision-Pro/ai_engine/yolov8_postprocess.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class LetterboxMeta:
    source_width: int
    source_height: int
    input_width: int = 640
    input_height: int = 640

    @property
    def scale(self) -> float:
        return min(self.input_width / self.source_width, self.input_height / self.source_height)

    @property
    def pad_x(self) -> float:
        return (self.input_width - self.source_width * self.scale) / 2

    @property
    def pad_y(self) -> float:
        return (self.input_height - self.source_height * self.scale) / 2


def dequantize_int8(values: Sequence[int], zero_point: int, scale: float) -> list[float]:
    return [(int(value) - zero_point) * scale for value in values]


def postprocess_yolov8_output(
    output: Sequence[Sequence[float]] | Sequence[float],
    source_width: int,
    source_height: int,
    input_width: int = 640,
    input_height: int = 640,
    num_classes: int = 7,
    confidence_threshold: float = 0.25,
    nms_iou_threshold: float = 0.45,
) -> list[list[float]]:
    channels = 4 + num_classes
    matrix = _as_channel_major(output, channels)
    meta = LetterboxMeta(
        source_width=source_width,
        source_height=source_height,
        input_width=input_width,
        input_height=input_height,
    )

    candidates: list[list[float]] = []
    anchor_count = len(matrix[0])
    for anchor_index in range(anchor_count):
        cx = matrix[0][anchor_index]
        cy = matrix[1][anchor_index]
        w = matrix[2][anchor_index]
        h = matrix[3][anchor_index]
        class_scores = [matrix[4 + class_id][anchor_index] for class_id in range(num_classes)]
        class_id, confidence = max(enumerate(class_scores), key=lambda item: item[1])
        if confidence < confidence_threshold:
            continue
        bbox = _xywh_to_source_xyxy(cx, cy, w, h, meta)
        if bbox is None:
            continue
        candidates.append([*bbox, confidence, float(class_id)])

    return nms(candidates, nms_iou_threshold)


def postprocess_yolov8_int8_output(
    output: Sequence[Sequence[int]] | Sequence[int],
    source_width: int,
    source_height: int,
    zero_point: int,
    scale: float,
    input_width: int = 640,
    input_height: int = 640,
    num_classes: int = 7,
    confidence_threshold: float = 0.25,
    nms_iou_threshold: float = 0.45,
) -> list[list[float]]:
    channels = 4 + num_classes
    matrix = _as_channel_major(output, channels)
    dequantized = [dequantize_int8(channel, zero_point, scale) for channel in matrix]
    return postprocess_yolov8_output(
        dequantized,
        source_width=source_width,
        source_height=source_height,
        input_width=input_width,
        input_height=input_height,
        num_classes=num_classes,
        confidence_threshold=confidence_threshold,
        nms_iou_threshold=nms_iou_threshold,
    )


def nms(boxes: list[list[float]], iou_threshold: float) -> list[list[float]]:
    selected: list[list[float]] = []
    remaining = sorted(boxes, key=lambda box: box[4], reverse=True)
    while remaining:
        current = remaining.pop(0)
        selected.append(current)
        remaining = [
            box
            for box in remaining
            if int(box[5]) != int(current[5]) or _iou_xyxy(box[:4], current[:4]) <= iou_threshold
        ]
    return selected


def _as_channel_major(
    output: Sequence[Sequence[float]] | Sequence[float] | Sequence[Sequence[int]] | Sequence[int],
    channels: int,
) -> list[list[float]]:
    if not output:
        return [[] for _ in range(channels)]

    first = output[0]  # type: ignore[index]
    if isinstance(first, (list, tuple)):
        rows = [list(row) for row in output]  # type: ignore[arg-type]
        if len(rows) == 1 and rows and isinstance(rows[0][0], (list, tuple)):
            rows = [list(row) for row in rows[0]]  # type: ignore[list-item]
        if len(rows) != channels:
            raise ValueError(f"YOLOv8 output must have {channels} channels, got {len(rows)}")
        return [[float(value) for value in row] for row in rows]

    flat = [float(value) for value in output]  # type: ignore[arg-type]
    if len(flat) % channels != 0:
        raise ValueError(f"flat YOLOv8 output length must be divisible by {channels}")
    anchor_count = len(flat) // channels
    return [flat[channel * anchor_count : (channel + 1) * anchor_count] for channel in range(channels)]


def _xywh_to_source_xyxy(cx: float, cy: float, w: float, h: float, meta: LetterboxMeta) -> list[float] | None:
    x1 = (cx - w / 2 - meta.pad_x) / meta.scale
    y1 = (cy - h / 2 - meta.pad_y) / meta.scale
    x2 = (cx + w / 2 - meta.pad_x) / meta.scale
    y2 = (cy + h / 2 - meta.pad_y) / meta.scale

    x1 = max(0.0, min(x1, float(meta.source_width - 1)))
    y1 = max(0.0, min(y1, float(meta.source_height - 1)))
    x2 = max(0.0, min(x2, float(meta.source_width - 1)))
    y2 = max(0.0, min(y2, float(meta.source_height - 1)))
    if x2 <= x1 or y2 <= y1:
        return None
    return [x1, y1, x2, y2]


def _iou_xyxy(a: Sequence[float], b: Sequence[float]) -> float:
    ax1, ay1, ax2, ay2 = a
    bx1, by1, bx2, by2 = b
    ix1 = max(ax1, bx1)
    iy1 = max(ay1, by1)
    ix2 = min(ax2, bx2)
    iy2 = min(ay2, by2)
    iw = max(ix2 - ix1, 0.0)
    ih = max(iy2 - iy1, 0.0)
    intersection = iw * ih
    area_a = max(ax2 - ax1, 0.0) * max(ay2 - ay1, 0.0)
    area_b = max(bx2 - bx1, 0.0) * max(by2 - by1, 0.0)
    union = area_a + area_b - intersection
    if union <= 0:
        return 0.0
    return intersection / union
